// @memgraph/orb/dist/events.js
import { Emitter } from './utils/emitter.utils';
export var OrbEventType;
(function (OrbEventType) {
    // Renderer events for drawing on canvas
    OrbEventType["RENDER_START"] = "render-start";
    OrbEventType["RENDER_END"] = "render-end";
    // Simulation (D3) events for setting up node positions
    OrbEventType["SIMULATION_START"] = "simulation-start";
    OrbEventType["SIMULATION_STEP"] = "simulation-step";
    OrbEventType["SIMULATION_END"] = "simulation-end";
    // Mouse events: click, hover, move
    OrbEventType["NODE_CLICK"] = "node-click";
    OrbEventType["NODE_HOVER"] = "node-hover";
    OrbEventType["EDGE_CLICK"] = "edge-click";
    OrbEventType["EDGE_HOVER"] = "edge-hover";
    OrbEventType["MOUSE_CLICK"] = "mouse-click";
    OrbEventType["MOUSE_MOVE"] = "mouse-move";
    // Zoom or pan (translate) change
    OrbEventType["TRANSFORM"] = "transform";
    // Mouse node drag events
    OrbEventType["NODE_DRAG_START"] = "node-drag-start";
    OrbEventType["NODE_DRAG"] = "node-drag";
    OrbEventType["NODE_DRAG_END"] = "node-drag-end";
    OrbEventType["NODE_RIGHT_CLICK"] = "node-right-click";
    OrbEventType["EDGE_RIGHT_CLICK"] = "edge-right-click";
    OrbEventType["MOUSE_RIGHT_CLICK"] = "mouse-right-click";
    // Double click events
    OrbEventType["NODE_DOUBLE_CLICK"] = "node-double-click";
    OrbEventType["EDGE_DOUBLE_CLICK"] = "edge-double-click";
    OrbEventType["MOUSE_DOUBLE_CLICK"] = "mouse-double-click";
})(OrbEventType || (OrbEventType = {}));
export class OrbEmitter extends Emitter {
}
