// @memgraph/orb/dist/utils/entity.utils.js
export class EntityState {
    constructor(definition) {
        this.ids = [];
        this.entityById = new Map();
        this.getId = definition.getId;
        this.sortBy = definition.sortBy;
    }
    getOne(id) {
        return this.entityById.get(id);
    }
    getMany(ids, options) {
        const entities = [];
        for (let i = 0; i < ids.length; i++) {
            const entity = this.getOne(ids[i]);
            if (entity === undefined) {
                continue;
            }
            if ((options === null || options === void 0 ? void 0 : options.filterBy) && !options.filterBy(entity)) {
                continue;
            }
            entities.push(entity);
        }
        if (this.sortBy) {
            entities.sort(this.sortBy);
        }
        return entities;
    }
    getAll(options) {
        const entities = [];
        for (let i = 0; i < this.ids.length; i++) {
            const entity = this.getOne(this.ids[i]);
            if (entity === undefined) {
                continue;
            }
            if ((options === null || options === void 0 ? void 0 : options.filterBy) && !options.filterBy(entity)) {
                continue;
            }
            entities.push(entity);
        }
        return entities;
    }
    setOne(entity) {
        const id = this.getId(entity);
        const isNewEntity = !this.entityById.has(id);
        this.entityById.set(id, entity);
        if (isNewEntity) {
            this.ids.push(id);
            this.sort();
        }
    }
    setMany(entities) {
        if (this.sortBy) {
            entities.sort(this.sortBy);
        }
        const newIds = [];
        for (let i = 0; i < entities.length; i++) {
            const entityId = this.getId(entities[i]);
            if (!this.entityById.has(entityId)) {
                newIds.push(entityId);
            }
            this.entityById.set(entityId, entities[i]);
        }
        this.ids = this.ids.concat(newIds);
        this.sort();
    }
    removeMany(ids) {
        const uniqueRemovedIds = new Set(ids);
        const newIds = [];
        for (let i = 0; i < this.ids.length; i++) {
            if (!uniqueRemovedIds.has(this.ids[i])) {
                newIds.push(this.ids[i]);
            }
        }
        this.ids = newIds;
        for (let i = 0; i < ids.length; i++) {
            this.entityById.delete(ids[i]);
        }
    }
    removeAll() {
        this.ids = [];
        this.entityById.clear();
    }
    sort() {
        if (!this.sortBy) {
            return;
        }
        this.ids.sort((id1, id2) => {
            // Typescript can't see the guard in the upper context
            if (!this.sortBy) {
                return 0;
            }
            const entity1 = this.getOne(id1);
            const entity2 = this.getOne(id2);
            // Should never happen
            if (entity1 === undefined || entity2 === undefined) {
                return 0;
            }
            return this.sortBy(entity1, entity2);
        });
    }
    get size() {
        return this.entityById.size;
    }
}
