// @memgraph/orb/dist/simulator/types/main-thread-simulator.js
import { SimulatorEventType } from '../shared';
import { Emitter } from '../../utils/emitter.utils';
import { D3SimulatorEngine, D3SimulatorEngineEventType, } from '../engine/d3-simulator-engine';
export class MainThreadSimulator extends Emitter {
    constructor() {
        super();
        this.simulator = new D3SimulatorEngine();
        this.simulator.on(D3SimulatorEngineEventType.SIMULATION_START, () => {
            this.emit(SimulatorEventType.SIMULATION_START, undefined);
        });
        this.simulator.on(D3SimulatorEngineEventType.SIMULATION_PROGRESS, (data) => {
            this.emit(SimulatorEventType.SIMULATION_PROGRESS, data);
        });
        this.simulator.on(D3SimulatorEngineEventType.SIMULATION_END, (data) => {
            this.emit(SimulatorEventType.SIMULATION_END, data);
        });
        this.simulator.on(D3SimulatorEngineEventType.NODE_DRAG, (data) => {
            this.emit(SimulatorEventType.NODE_DRAG_END, data);
        });
        this.simulator.on(D3SimulatorEngineEventType.TICK, (data) => {
            this.emit(SimulatorEventType.NODE_DRAG, data);
        });
        this.simulator.on(D3SimulatorEngineEventType.END, (data) => {
            this.emit(SimulatorEventType.NODE_DRAG_END, data);
        });
        this.simulator.on(D3SimulatorEngineEventType.SETTINGS_UPDATE, (data) => {
            this.emit(SimulatorEventType.SETTINGS_UPDATE, data);
        });
    }
    setData(nodes, edges) {
        this.simulator.setData({ nodes, edges });
    }
    addData(nodes, edges) {
        this.simulator.addData({ nodes, edges });
    }
    updateData(nodes, edges) {
        this.simulator.updateData({ nodes, edges });
    }
    clearData() {
        this.simulator.clearData();
    }
    simulate() {
        this.simulator.simulate();
    }
    activateSimulation() {
        this.simulator.activateSimulation();
    }
    startSimulation(nodes, edges) {
        this.simulator.startSimulation({ nodes, edges });
    }
    updateSimulation(nodes, edges) {
        this.simulator.updateSimulation({ nodes, edges });
    }
    stopSimulation() {
        this.simulator.stopSimulation();
    }
    startDragNode() {
        this.simulator.startDragNode();
    }
    dragNode(nodeId, position) {
        this.simulator.dragNode(Object.assign({ id: nodeId }, position));
    }
    endDragNode(nodeId) {
        this.simulator.endDragNode({ id: nodeId });
    }
    fixNodes(nodes) {
        this.simulator.fixNodes(nodes);
    }
    releaseNodes(nodes) {
        this.simulator.releaseNodes(nodes);
    }
    setSettings(settings) {
        this.simulator.setSettings(settings);
    }
    terminate() {
        this.simulator.removeAllListeners();
        this.removeAllListeners();
        // Do nothing
    }
}
