// @memgraph/orb/dist/models/graph.js
import { NodeFactory } from './node';
import { EdgeFactory } from './edge';
import { ImageHandler } from '../services/images';
import { getEdgeOffsets } from './topology';
import { EntityState } from '../utils/entity.utils';
export class Graph {
    constructor(data, settings) {
        var _a, _b;
        this._nodes = new EntityState({
            getId: (node) => node.id,
            sortBy: (node1, node2) => { var _a, _b; return ((_a = node1.style.zIndex) !== null && _a !== void 0 ? _a : 0) - ((_b = node2.style.zIndex) !== null && _b !== void 0 ? _b : 0); },
        });
        this._edges = new EntityState({
            getId: (edge) => edge.id,
            sortBy: (edge1, edge2) => { var _a, _b; return ((_a = edge1.style.zIndex) !== null && _a !== void 0 ? _a : 0) - ((_b = edge2.style.zIndex) !== null && _b !== void 0 ? _b : 0); },
        });
        this._onLoadedImages = settings === null || settings === void 0 ? void 0 : settings.onLoadedImages;
        const nodes = (_a = data === null || data === void 0 ? void 0 : data.nodes) !== null && _a !== void 0 ? _a : [];
        const edges = (_b = data === null || data === void 0 ? void 0 : data.edges) !== null && _b !== void 0 ? _b : [];
        this.setup({ nodes, edges });
    }
    /**
     * Returns a list of nodes.
     *
     * @param {INodeFilter} filterBy Filter function for nodes
     * @return {INode[]} List of nodes
     */
    getNodes(filterBy) {
        return this._nodes.getAll({ filterBy });
    }
    /**
     * Returns a list of edges.
     *
     * @param {IEdgeFilter} filterBy Filter function for edges
     * @return {IEdge[]} List of edges
     */
    getEdges(filterBy) {
        return this._edges.getAll({ filterBy });
    }
    /**
     * Returns the total node count.
     *
     * @return {number} Total node count
     */
    getNodeCount() {
        return this._nodes.size;
    }
    /**
     * Returns the total edge count.
     *
     * @return {number} Total edge count
     */
    getEdgeCount() {
        return this._edges.size;
    }
    /**
     * Returns node by node id.
     *
     * @param {any} id Node id
     * @return {Node | undefined} Node or undefined
     */
    getNodeById(id) {
        return this._nodes.getOne(id);
    }
    /**
     * Returns edge by edge id.
     *
     * @param {any} id Edge id
     * @return {IEdge | undefined} Edge or undefined
     */
    getEdgeById(id) {
        return this._edges.getOne(id);
    }
    /**
     * Returns a list of current node positions (x, y).
     *
     * @return {INodePosition[]} List of node positions
     */
    getNodePositions() {
        const nodes = this.getNodes();
        const positions = new Array(nodes.length);
        for (let i = 0; i < nodes.length; i++) {
            positions[i] = nodes[i].position;
        }
        return positions;
    }
    /**
     * Sets new node positions (x, y).
     *
     * @param {INodePosition} positions Node positions
     */
    setNodePositions(positions) {
        for (let i = 0; i < positions.length; i++) {
            const node = this._nodes.getOne(positions[i].id);
            if (node) {
                node.position = positions[i];
            }
        }
    }
    /**
     * Returns a list of current edge positions. Edge positions do not have
     * (x, y) but a link to the source and target node ids.
     *
     * @return {IEdgePosition[]} List of edge positions
     */
    getEdgePositions() {
        const edges = this.getEdges();
        const positions = new Array(edges.length);
        for (let i = 0; i < edges.length; i++) {
            positions[i] = edges[i].position;
        }
        return positions;
    }
    /**
     * Sets default style to new nodes and edges. The applied style will be used
     * for all future nodes and edges added with `.merge` function.
     *
     * @param {IGraphStyle} style Style definition
     */
    setDefaultStyle(style) {
        this._defaultStyle = style;
        this._applyStyle();
    }
    setup(data) {
        var _a, _b;
        this._nodes.removeAll();
        this._edges.removeAll();
        const nodes = (_a = data === null || data === void 0 ? void 0 : data.nodes) !== null && _a !== void 0 ? _a : [];
        const edges = (_b = data === null || data === void 0 ? void 0 : data.edges) !== null && _b !== void 0 ? _b : [];
        this._insertNodes(nodes);
        this._insertEdges(edges);
        this._applyEdgeOffsets();
        this._applyStyle();
    }
    clearPositions() {
        const nodes = this.getNodes();
        for (let i = 0; i < nodes.length; i++) {
            nodes[i].clearPosition();
        }
    }
    merge(data) {
        var _a, _b;
        const nodes = (_a = data.nodes) !== null && _a !== void 0 ? _a : [];
        const edges = (_b = data.edges) !== null && _b !== void 0 ? _b : [];
        this._upsertNodes(nodes);
        this._upsertEdges(edges);
        this._applyEdgeOffsets();
        this._applyStyle();
    }
    remove(data) {
        var _a, _b;
        const nodeIds = (_a = data.nodeIds) !== null && _a !== void 0 ? _a : [];
        const edgeIds = (_b = data.edgeIds) !== null && _b !== void 0 ? _b : [];
        this._removeNodes(nodeIds);
        this._removeEdges(edgeIds);
        this._applyEdgeOffsets();
        this._applyStyle();
    }
    isEqual(graph) {
        if (this.getNodeCount() !== graph.getNodeCount()) {
            return false;
        }
        if (this.getEdgeCount() !== graph.getEdgeCount()) {
            return false;
        }
        const nodes = this.getNodes();
        for (let i = 0; i < nodes.length; i++) {
            if (!graph.getNodeById(nodes[i].id)) {
                return false;
            }
        }
        const edges = this.getEdges();
        for (let i = 0; i < edges.length; i++) {
            if (!graph.getEdgeById(edges[i].id)) {
                return false;
            }
        }
        return true;
    }
    getBoundingBox() {
        const nodes = this.getNodes();
        const minPoint = { x: 0, y: 0 };
        const maxPoint = { x: 0, y: 0 };
        for (let i = 0; i < nodes.length; i++) {
            const { x, y } = nodes[i].getCenter();
            if (x === undefined || y === undefined) {
                continue;
            }
            const size = nodes[i].getBorderedRadius();
            if (i === 0) {
                minPoint.x = x - size;
                maxPoint.x = x + size;
                minPoint.y = y - size;
                maxPoint.y = y + size;
                continue;
            }
            if (x + size > maxPoint.x) {
                maxPoint.x = x + size;
            }
            if (x - size < minPoint.x) {
                minPoint.x = x - size;
            }
            if (y + size > maxPoint.y) {
                maxPoint.y = y + size;
            }
            if (y - size < minPoint.y) {
                minPoint.y = y - size;
            }
        }
        return {
            x: minPoint.x,
            y: minPoint.y,
            width: Math.abs(maxPoint.x - minPoint.x),
            height: Math.abs(maxPoint.y - minPoint.y),
        };
    }
    getNearestNode(point) {
        // Reverse is needed to check from the top drawn to the bottom drawn node
        const nodes = this.getNodes();
        for (let i = nodes.length - 1; i >= 0; i--) {
            if (nodes[i].includesPoint(point)) {
                return nodes[i];
            }
        }
    }
    getNearestEdge(point, minDistance = 3) {
        let nearestEdge;
        let nearestDistance = minDistance;
        const edges = this.getEdges();
        for (let i = 0; i < edges.length; i++) {
            const distance = edges[i].getDistance(point);
            if (distance <= nearestDistance) {
                nearestDistance = distance;
                nearestEdge = edges[i];
            }
        }
        return nearestEdge;
    }
    _insertNodes(nodes) {
        const newNodes = new Array(nodes.length);
        for (let i = 0; i < nodes.length; i++) {
            newNodes[i] = NodeFactory.create({ data: nodes[i] }, { onLoadedImage: () => { var _a; return (_a = this._onLoadedImages) === null || _a === void 0 ? void 0 : _a.call(this); } });
        }
        this._nodes.setMany(newNodes);
    }
    _insertEdges(edges) {
        const newEdges = [];
        for (let i = 0; i < edges.length; i++) {
            const startNode = this.getNodeById(edges[i].start);
            const endNode = this.getNodeById(edges[i].end);
            if (startNode && endNode) {
                newEdges.push(EdgeFactory.create({
                    data: edges[i],
                    startNode,
                    endNode,
                }));
            }
        }
        this._edges.setMany(newEdges);
    }
    _upsertNodes(nodes) {
        const newNodes = [];
        for (let i = 0; i < nodes.length; i++) {
            const existingNode = this.getNodeById(nodes[i].id);
            if (existingNode) {
                existingNode.data = nodes[i];
                continue;
            }
            newNodes.push(NodeFactory.create({ data: nodes[i] }, { onLoadedImage: () => { var _a; return (_a = this._onLoadedImages) === null || _a === void 0 ? void 0 : _a.call(this); } }));
        }
        this._nodes.setMany(newNodes);
    }
    _upsertEdges(edges) {
        const newEdges = [];
        const removedEdgeIds = [];
        for (let i = 0; i < edges.length; i++) {
            const newEdgeData = edges[i];
            const existingEdge = this.getEdgeById(newEdgeData.id);
            // New edge
            if (!existingEdge) {
                const startNode = this.getNodeById(newEdgeData.start);
                const endNode = this.getNodeById(newEdgeData.end);
                if (startNode && endNode) {
                    const edge = EdgeFactory.create({
                        data: newEdgeData,
                        startNode,
                        endNode,
                    });
                    newEdges.push(edge);
                }
                continue;
            }
            // The connection of the edge stays the same, but the data has changed
            if (existingEdge.start === newEdgeData.start && existingEdge.end === newEdgeData.end) {
                existingEdge.data = newEdgeData;
                continue;
            }
            // Edge connection (start or end node) has changed
            existingEdge.startNode.removeEdge(existingEdge);
            existingEdge.endNode.removeEdge(existingEdge);
            const startNode = this.getNodeById(newEdgeData.start);
            const endNode = this.getNodeById(newEdgeData.end);
            if (!startNode || !endNode) {
                removedEdgeIds.push(existingEdge.id);
                continue;
            }
            const edge = EdgeFactory.create({
                data: newEdgeData,
                offset: existingEdge.offset,
                startNode,
                endNode,
            });
            edge.state = existingEdge.state;
            edge.style = existingEdge.style;
            newEdges.push(edge);
        }
        this._edges.setMany(newEdges);
        this._edges.removeMany(removedEdgeIds);
    }
    _removeNodes(nodeIds) {
        const removedNodeIds = [];
        const removedEdgeIds = [];
        for (let i = 0; i < nodeIds.length; i++) {
            const node = this.getNodeById(nodeIds[i]);
            if (!node) {
                continue;
            }
            const edges = node.getEdges();
            for (let i = 0; i < edges.length; i++) {
                const edge = edges[i];
                edge.startNode.removeEdge(edge);
                edge.endNode.removeEdge(edge);
                removedEdgeIds.push(edge.id);
            }
            removedNodeIds.push(node.id);
        }
        this._nodes.removeMany(removedNodeIds);
        this._edges.removeMany(removedEdgeIds);
    }
    _removeEdges(edgeIds) {
        const removedEdgeIds = [];
        for (let i = 0; i < edgeIds.length; i++) {
            const edge = this.getEdgeById(edgeIds[i]);
            if (!edge) {
                continue;
            }
            edge.startNode.removeEdge(edge);
            edge.endNode.removeEdge(edge);
            removedEdgeIds.push(edge.id);
        }
        this._edges.removeMany(removedEdgeIds);
    }
    _applyEdgeOffsets() {
        const graphEdges = this.getEdges();
        const edgeOffsets = getEdgeOffsets(graphEdges);
        const updatedEdges = new Array(edgeOffsets.length);
        for (let i = 0; i < edgeOffsets.length; i++) {
            const edge = graphEdges[i];
            const edgeOffset = edgeOffsets[i];
            updatedEdges[i] = EdgeFactory.copy(edge, { offset: edgeOffset });
        }
        this._edges.setMany(updatedEdges);
    }
    _applyStyle() {
        var _a, _b;
        const styleImageUrls = new Set();
        if ((_a = this._defaultStyle) === null || _a === void 0 ? void 0 : _a.getNodeStyle) {
            const newNodes = this.getNodes();
            for (let i = 0; i < newNodes.length; i++) {
                if (newNodes[i].hasStyle()) {
                    continue;
                }
                const style = this._defaultStyle.getNodeStyle(newNodes[i]);
                if (style) {
                    newNodes[i].style = style;
                    // TODO Add these checks to node property setup
                    if (style.imageUrl) {
                        styleImageUrls.add(style.imageUrl);
                    }
                    if (style.imageUrlSelected) {
                        styleImageUrls.add(style.imageUrlSelected);
                    }
                }
            }
        }
        if ((_b = this._defaultStyle) === null || _b === void 0 ? void 0 : _b.getEdgeStyle) {
            const newEdges = this.getEdges();
            for (let i = 0; i < newEdges.length; i++) {
                if (newEdges[i].hasStyle()) {
                    continue;
                }
                const style = this._defaultStyle.getEdgeStyle(newEdges[i]);
                if (style) {
                    newEdges[i].style = style;
                }
            }
        }
        if (styleImageUrls.size) {
            ImageHandler.getInstance().loadImages(Array.from(styleImageUrls), () => {
                var _a;
                (_a = this._onLoadedImages) === null || _a === void 0 ? void 0 : _a.call(this);
            });
        }
        // Sort because of the possibility that style properties changed
        this._nodes.sort();
        this._edges.sort();
    }
}
