// @memgraph/orb/dist/models/node.js
import { isPointInRectangle } from '../common';
import { ImageHandler } from '../services/images';
import { GraphObjectState } from './state';
export var NodeShapeType;
(function (NodeShapeType) {
    NodeShapeType["CIRCLE"] = "circle";
    NodeShapeType["DOT"] = "dot";
    NodeShapeType["SQUARE"] = "square";
    NodeShapeType["DIAMOND"] = "diamond";
    NodeShapeType["TRIANGLE"] = "triangle";
    NodeShapeType["TRIANGLE_DOWN"] = "triangleDown";
    NodeShapeType["STAR"] = "star";
    NodeShapeType["HEXAGON"] = "hexagon";
})(NodeShapeType || (NodeShapeType = {}));
export class NodeFactory {
    static create(data, settings) {
        return new Node(data, settings);
    }
}
export const isNode = (obj) => {
    return obj instanceof Node;
};
export class Node {
    constructor(data, settings) {
        this.style = {};
        this.state = GraphObjectState.NONE;
        this._inEdgesById = {};
        this._outEdgesById = {};
        this.id = data.data.id;
        this.data = data.data;
        this.position = { id: this.id };
        this._onLoadedImage = settings === null || settings === void 0 ? void 0 : settings.onLoadedImage;
    }
    clearPosition() {
        this.position.x = undefined;
        this.position.y = undefined;
    }
    getCenter() {
        // This should not be called in the render because nodes without position will be
        // filtered out
        if (this.position.x === undefined || this.position.y === undefined) {
            return { x: 0, y: 0 };
        }
        return { x: this.position.x, y: this.position.y };
    }
    getRadius() {
        var _a;
        return (_a = this.style.size) !== null && _a !== void 0 ? _a : 0;
    }
    getBorderedRadius() {
        return this.getRadius() + this.getBorderWidth() / 2;
    }
    getBoundingBox() {
        const center = this.getCenter();
        const radius = this.getBorderedRadius();
        return {
            x: center.x - radius,
            y: center.y - radius,
            width: radius * 2,
            height: radius * 2,
        };
    }
    getInEdges() {
        return Object.values(this._inEdgesById);
    }
    getOutEdges() {
        return Object.values(this._outEdgesById);
    }
    getEdges() {
        const edgeById = {};
        const outEdges = this.getOutEdges();
        for (let i = 0; i < outEdges.length; i++) {
            const outEdge = outEdges[i];
            edgeById[outEdge.id] = outEdge;
        }
        const inEdges = this.getInEdges();
        for (let i = 0; i < inEdges.length; i++) {
            const inEdge = inEdges[i];
            edgeById[inEdge.id] = inEdge;
        }
        return Object.values(edgeById);
    }
    getAdjacentNodes() {
        const adjacentNodeById = {};
        const outEdges = this.getOutEdges();
        for (let i = 0; i < outEdges.length; i++) {
            const adjacentNode = outEdges[i].endNode;
            if (adjacentNode) {
                adjacentNodeById[adjacentNode.id] = adjacentNode;
            }
        }
        const inEdges = this.getInEdges();
        for (let i = 0; i < inEdges.length; i++) {
            const adjacentNode = inEdges[i].startNode;
            if (adjacentNode) {
                adjacentNodeById[adjacentNode.id] = adjacentNode;
            }
        }
        return Object.values(adjacentNodeById);
    }
    hasStyle() {
        return this.style && Object.keys(this.style).length > 0;
    }
    addEdge(edge) {
        if (edge.start === this.id) {
            this._outEdgesById[edge.id] = edge;
        }
        if (edge.end === this.id) {
            this._inEdgesById[edge.id] = edge;
        }
    }
    removeEdge(edge) {
        delete this._outEdgesById[edge.id];
        delete this._inEdgesById[edge.id];
    }
    isSelected() {
        return this.state === GraphObjectState.SELECTED;
    }
    isHovered() {
        return this.state === GraphObjectState.HOVERED;
    }
    clearState() {
        this.state = GraphObjectState.NONE;
    }
    getDistanceToBorder() {
        // TODO: Add "getDistanceToBorder(angle: number)" for each node shape type because this covers only circles
        return this.getBorderedRadius();
    }
    includesPoint(point) {
        const isInBoundingBox = this._isPointInBoundingBox(point);
        if (!isInBoundingBox) {
            return false;
        }
        // For square type, we don't need to check the circle
        if (this.style.shape === NodeShapeType.SQUARE) {
            return isInBoundingBox;
        }
        // TODO: Add better "includePoint" checks for stars, triangles, hexagons, etc.
        const center = this.getCenter();
        const borderedRadius = this.getBorderedRadius();
        const dx = point.x - center.x;
        const dy = point.y - center.y;
        return Math.sqrt(dx * dx + dy * dy) <= borderedRadius;
    }
    hasShadow() {
        var _a, _b, _c;
        return (((_a = this.style.shadowSize) !== null && _a !== void 0 ? _a : 0) > 0 || ((_b = this.style.shadowOffsetX) !== null && _b !== void 0 ? _b : 0) > 0 || ((_c = this.style.shadowOffsetY) !== null && _c !== void 0 ? _c : 0) > 0);
    }
    hasBorder() {
        var _a, _b;
        const hasBorderWidth = ((_a = this.style.borderWidth) !== null && _a !== void 0 ? _a : 0) > 0;
        const hasBorderWidthSelected = ((_b = this.style.borderWidthSelected) !== null && _b !== void 0 ? _b : 0) > 0;
        return hasBorderWidth || (this.isSelected() && hasBorderWidthSelected);
    }
    getLabel() {
        return this.style.label;
    }
    getColor() {
        let color = undefined;
        if (this.style.color) {
            color = this.style.color;
        }
        if (this.isHovered() && this.style.colorHover) {
            color = this.style.colorHover;
        }
        if (this.isSelected() && this.style.colorSelected) {
            color = this.style.colorSelected;
        }
        return color;
    }
    getBorderWidth() {
        let borderWidth = 0;
        if (this.style.borderWidth && this.style.borderWidth > 0) {
            borderWidth = this.style.borderWidth;
        }
        if (this.isSelected() && this.style.borderWidthSelected && this.style.borderWidthSelected > 0) {
            borderWidth = this.style.borderWidthSelected;
        }
        return borderWidth;
    }
    getBorderColor() {
        if (!this.hasBorder()) {
            return undefined;
        }
        let borderColor = undefined;
        if (this.style.borderColor) {
            borderColor = this.style.borderColor;
        }
        if (this.isHovered() && this.style.borderColorHover) {
            borderColor = this.style.borderColorHover;
        }
        if (this.isSelected() && this.style.borderColorSelected) {
            borderColor = this.style.borderColorSelected.toString();
        }
        return borderColor;
    }
    getBackgroundImage() {
        var _a;
        if (((_a = this.style.size) !== null && _a !== void 0 ? _a : 0) <= 0) {
            return;
        }
        let imageUrl;
        if (this.style.imageUrl) {
            imageUrl = this.style.imageUrl;
        }
        if (this.isSelected() && this.style.imageUrlSelected) {
            imageUrl = this.style.imageUrlSelected;
        }
        if (!imageUrl) {
            return;
        }
        const image = ImageHandler.getInstance().getImage(imageUrl);
        if (image) {
            return image;
        }
        return ImageHandler.getInstance().loadImage(imageUrl, (error) => {
            var _a;
            if (!error) {
                (_a = this._onLoadedImage) === null || _a === void 0 ? void 0 : _a.call(this);
            }
        });
    }
    _isPointInBoundingBox(point) {
        return isPointInRectangle(this.getBoundingBox(), point);
    }
}
