// @memgraph/orb/dist/renderer/webgl/webgl-renderer.js
import { zoomIdentity } from 'd3-zoom';
import { Emitter } from '../../utils/emitter.utils';
import { DEFAULT_RENDERER_HEIGHT, DEFAULT_RENDERER_SETTINGS, DEFAULT_RENDERER_WIDTH, } from '../shared';
import { copyObject } from '../../utils/object.utils';
export class WebGLRenderer extends Emitter {
    constructor(context, settings) {
        super();
        this._context = context;
        console.log('context', this._context);
        this.width = DEFAULT_RENDERER_WIDTH;
        this.height = DEFAULT_RENDERER_HEIGHT;
        this.transform = zoomIdentity;
        this._settings = Object.assign(Object.assign({}, DEFAULT_RENDERER_SETTINGS), settings);
    }
    get isInitiallyRendered() {
        throw new Error('Method not implemented.');
    }
    getSettings() {
        return copyObject(this._settings);
    }
    setSettings(settings) {
        this._settings = Object.assign(Object.assign({}, this._settings), settings);
    }
    render(graph) {
        console.log('graph:', graph);
        throw new Error('Method not implemented.');
    }
    reset() {
        throw new Error('Method not implemented.');
    }
    getFitZoomTransform(graph) {
        console.log('graph:', graph);
        throw new Error('Method not implemented.');
    }
    getSimulationPosition(canvasPoint) {
        console.log('canvasPoint:', canvasPoint);
        throw new Error('Method not implemented.');
    }
    getSimulationViewRectangle() {
        throw new Error('Method not implemented.');
    }
    translateOriginToCenter() {
        throw new Error('Method not implemented.');
    }
}
