// @memgraph/orb/dist/simulator/types/web-worker-simulator/simulator.js
import { SimulatorEventType } from '../../shared';
import { WorkerInputType } from './message/worker-input';
import { WorkerOutputType } from './message/worker-output';
import { Emitter } from '../../../utils/emitter.utils';
export class WebWorkerSimulator extends Emitter {
    constructor() {
        super();
        this.worker = new Worker(new URL(
        /* webpackChunkName: 'process.worker' */
        './process.worker', import.meta.url), { type: 'module' });
        this.worker.onmessage = ({ data }) => {
            switch (data.type) {
                case WorkerOutputType.SIMULATION_START: {
                    this.emit(SimulatorEventType.SIMULATION_START, undefined);
                    break;
                }
                case WorkerOutputType.SIMULATION_PROGRESS: {
                    this.emit(SimulatorEventType.SIMULATION_PROGRESS, data.data);
                    break;
                }
                case WorkerOutputType.SIMULATION_END: {
                    this.emit(SimulatorEventType.SIMULATION_END, data.data);
                    break;
                }
                case WorkerOutputType.NODE_DRAG: {
                    this.emit(SimulatorEventType.NODE_DRAG, data.data);
                    break;
                }
                case WorkerOutputType.NODE_DRAG_END: {
                    this.emit(SimulatorEventType.NODE_DRAG_END, data.data);
                    break;
                }
                case WorkerOutputType.SETTINGS_UPDATE: {
                    this.emit(SimulatorEventType.SETTINGS_UPDATE, data.data);
                    break;
                }
            }
        };
    }
    setData(nodes, edges) {
        this.emitToWorker({ type: WorkerInputType.SetData, data: { nodes, edges } });
    }
    addData(nodes, edges) {
        this.emitToWorker({ type: WorkerInputType.AddData, data: { nodes, edges } });
    }
    updateData(nodes, edges) {
        this.emitToWorker({ type: WorkerInputType.UpdateData, data: { nodes, edges } });
    }
    clearData() {
        this.emitToWorker({ type: WorkerInputType.ClearData });
    }
    simulate() {
        this.emitToWorker({ type: WorkerInputType.Simulate });
    }
    activateSimulation() {
        this.emitToWorker({ type: WorkerInputType.ActivateSimulation });
    }
    startSimulation(nodes, edges) {
        this.emitToWorker({ type: WorkerInputType.StartSimulation, data: { nodes, edges } });
    }
    updateSimulation(nodes, edges) {
        this.emitToWorker({ type: WorkerInputType.UpdateSimulation, data: { nodes, edges } });
    }
    stopSimulation() {
        this.emitToWorker({ type: WorkerInputType.StopSimulation });
    }
    startDragNode() {
        this.emitToWorker({ type: WorkerInputType.StartDragNode });
    }
    dragNode(nodeId, position) {
        this.emitToWorker({ type: WorkerInputType.DragNode, data: Object.assign({ id: nodeId }, position) });
    }
    endDragNode(nodeId) {
        this.emitToWorker({ type: WorkerInputType.EndDragNode, data: { id: nodeId } });
    }
    fixNodes(nodes) {
        this.emitToWorker({ type: WorkerInputType.FixNodes, data: { nodes } });
    }
    releaseNodes(nodes) {
        this.emitToWorker({ type: WorkerInputType.ReleaseNodes, data: { nodes } });
    }
    setSettings(settings) {
        this.emitToWorker({ type: WorkerInputType.SetSettings, data: settings });
    }
    terminate() {
        this.worker.terminate();
        this.removeAllListeners();
    }
    emitToWorker(message) {
        this.worker.postMessage(message);
    }
}
